### PR TITLE
FuzzyHelper refactoring + Ai Map Object Evaluation improvements

### PR DESCRIPTION
**Pull request by dydzio0614**
- [x] Split FuzzyHelper logic into extendable goal engine classes
- [x] Use smart pointers for FuzzyLite input variables (ABANDONED - should not be done)
- [x] Create elementar version of GetObj goal
- [x] Decide how to handle creature dwelling AI values and implement solution
- [x] Split Fuzzy.h/cpp files

### CODE REVIEW

File: AI/VCAI/Fuzzy.h [hunk 1]
```diff
@@ -27,42 +27,52 @@ class engineBase
...
+};
+
+class EvalVisitTile : public HeroMovementGoalEngine
+{
+public:
+	~EvalVisitTile();
+};
+
+class EvalWanderTargetObject : public HeroMovementGoalEngine //designed for use with VCAI::wander()
+{
```
Comment (DjWarmonger) [inline AI/VCAI/Fuzzy.h]: That's right!

File: AI/VCAI/Fuzzy.cpp [hunk 2]
```diff
@@ -373,142 +373,93 @@ float FuzzyHelper::evaluate(Goals::RecruitHero & g)
...
+		heroStrength->addTerm(new fl::Ramp("LOW", 0.2, 0));
+		heroStrength->addTerm(new fl::Triangle("MEDIUM", 0.2, 0.8));
+		heroStrength->addTerm(new fl::Ramp("HIGH", 0.5, 1));
+		heroStrength->setRange(0.0, 1.0);
 
-		vt.turnDistance->addTerm(new fl::Ramp("SMALL", 0.5, 0));
-		vt.turnDistance->addTerm(new fl::Triangle("MEDIUM", 0.1, 0.8));
-		vt.turnDistance->addTerm(new fl::Ramp("LONG", 0.5, 3));
-		vt.turnDistance->setRange(0.0, 3.0);
+		turnDistance->addTerm(new fl::Ramp("SMALL", 0.5, 0));
```
Comment (alexvins) [inline AI/VCAI/Fuzzy.cpp]: Should be `SHORT`

Comment (alexvins) [reply AI/VCAI/Fuzzy.cpp]: Also if we use MEDIUM term here, why not use LOW and HIGH too? And input variable configuration Ramp-Triangle-Ramp should factored out to function or even to subclass of `fl::InputVariable`

File: AI/VCAI/Fuzzy.cpp [hunk 3]
```diff
@@ -373,142 +373,93 @@ float FuzzyHelper::evaluate(Goals::RecruitHero & g)
...
+		strengthRatio->addTerm(new fl::Ramp("HIGH", SAFE_ATTACK_CONSTANT, SAFE_ATTACK_CONSTANT * 3));
+		strengthRatio->setRange(0, SAFE_ATTACK_CONSTANT * 3);
 
 		//strength compared to our main hero
-		vt.heroStrength->addTerm(new fl::Ramp("LOW", 0.2, 0));
-		vt.heroStrength->addTerm(new fl::Triangle("MEDIUM", 0.2, 0.8));
-		vt.heroStrength->addTerm(new fl::Ramp("HIGH", 0.5, 1));
-		vt.heroStrength->setRange(0.0, 1.0);
+		heroStrength->addTerm(new fl::Ramp("LOW", 0.2, 0));
+		heroStrength->addTerm(new fl::Triangle("MEDIUM", 0.2, 0.8));
```
Comment (alexvins) [inline AI/VCAI/Fuzzy.cpp]: If input is exactly 0.2, no term will be actually active

Comment (DjWarmonger) [reply AI/VCAI/Fuzzy.cpp]: Terms must cover entire range of values, and at least one rule must trigger for every combination.

Comment (alexvins) [reply AI/VCAI/Fuzzy.cpp]: Terms of input variables must overlap, it this case at 0.2 no rule will be triggered (membership function is zero for all of them)

File: AI/VCAI/Fuzzy.cpp [hunk 4]
```diff
@@ -373,142 +373,93 @@ float FuzzyHelper::evaluate(Goals::RecruitHero & g)
...
+		heroStrength->setRange(0.0, 1.0);
 
-		vt.turnDistance->addTerm(new fl::Ramp("SMALL", 0.5, 0));
-		vt.turnDistance->addTerm(new fl::Triangle("MEDIUM", 0.1, 0.8));
-		vt.turnDistance->addTerm(new fl::Ramp("LONG", 0.5, 3));
-		vt.turnDistance->setRange(0.0, 3.0);
+		turnDistance->addTerm(new fl::Ramp("SMALL", 0.5, 0));
+		turnDistance->addTerm(new fl::Triangle("MEDIUM", 0.1, 0.8));
+		turnDistance->addTerm(new fl::Ramp("LONG", 0.5, 3));
+		turnDistance->setRange(0.0, 3.0);
```
Comment (alexvins) [inline AI/VCAI/Fuzzy.cpp]: All terms have default height (1.0) but variables have custom ranges, I doubt that this is intended.

**Comment by dydzio0614** [comment]
About dwellings... I tried, but with current objects code I was unable to calculate dwelling values. Creature data for particular dwelling type is apparently initialized at object instantiation and I found no way to handle it properly via VLC->objtypeh or VLC->creh

**Comment by DjWarmonger** [comment]
But what is the problem? For every dwelling object we can determine the creature type(s) it gives. AI can already buy creatures from dwellings, so it can also know their parameters.

**Comment by dydzio0614** [comment]
We cannot determine the creature it gives for mods. We can read json node for creature name, but creature handler cannot resolve creature by name alone, because it also needs creature namespace as 2nd parameter - for non-mod creatures it is "core".

**Comment by DjWarmonger** [comment]
Don't read any json here lol, just use actual dwelling properties from *obj.
Like here: https://github.com/vcmi/vcmi/blob/develop/AI/VCAI/Goals.cpp#L1545

**Comment by dydzio0614** [comment]
The code you provided works on CGObjectInstance, map objects evaluator works on TObjectTypeHandler and is not tied to objects on current map.

**Comment by DjWarmonger** [comment]
So it's bad design on your side :P VCAI always evaluates concrete objects of CGObjectInstance * it sees in a map.

**Comment by dydzio0614** [comment]
Debatable, I find this logic pretty straightforward, human player also knows objects that do not exist on particular map :P

**Comment by DjWarmonger** [comment]
Well yes, but VCAI has no use of objects that do not exist on a map. Especially not for fuzzy evaluation. You are probably confusing FindObj with GetObj, which could be renamed to VisitObj and for all purposes is equal to VisitTile.

**Comment by dydzio0614** [comment]
Until 2 days ago I thought it is possible to get all object data without creating CGObjectInstance and this is the annoying design part IMO

**Comment by DjWarmonger** [comment]
Uhh, many comments in VCAI say "TODO" for that matter :P

**Comment by DjWarmonger** [comment]
Either way, I don't see why MapObjectsEvaluator couldn't convert CGObjectInstance to ID and subID, this is trivial to implement.

**Comment by dydzio0614** [comment]
MapObjectEvaluator state is persistent between games as far as I remember and constructor adding objects to database is not called again when starting new map with new objects

**Comment by dydzio0614** [comment]
I may try to create mock objectinstances for all dwellings but I do not guarantee good performance of that solution

**Comment by DjWarmonger** [comment]
Can stay like that if you wish to, but then again I made this graphs for a purpose: https://wiki.vcmi.eu/File:VCAI_Target.png

**Comment by DjWarmonger** [comment]
What on earth are you talking about, just add a function that takes CGObjectInstance * as an argument

**Comment by dydzio0614** [comment]
Okay, this branch is done. Needs to be tested before merge.

**Comment by DjWarmonger** [comment]
Error from console:
```
FindWanderTarget: [syntax error] consequent expected output variable, but found <value>
{at \ai\fuzzylite\fuzzylite\src\rule\consequent.cpp::fl::Consequent::load() [line:182]}
```

**Comment by DjWarmonger** [comment]
Also getting crash every time, in different situations:
`Cannot get object with id 5135. Object was removed`

**Comment by DjWarmonger** [comment]
Okay I think we can merge this one. I already have a pack of bugfixes based on this branch.

**Comment by dydzio0614** [comment]
Yeah, I agree.

**Pull request merged**